src/create-template-orders.js

```diff
@@ -20,20 +33,263 @@ async function _fetchLastStartTimestamp() {
+  const maxRetry = 20
+  while (true)
+    try {
+      await apiRoot
+        .orders()
+        .withId({ ID: templateOrder.id })
+        .post({
+          body: {
+            actions: addPaymentActions,
+            version,
+          },
+        })
+        .execute()
+      break
+    } catch (err) {
+      retryCount += 1
+      const currentVersion = err.body.errors[0].currentVersion
+      if (retryCount > maxRetry) {
+        const retryMessage =
+          'Got a concurrent modification error' +
+          ` when creating template order with number "${templateOrder.orderNumber}".` +
+          ` Version tried "${version}",` +
+          ` currentVersion: "${currentVersion}".`
+        throw new VError(
+          err,
+          `${retryMessage} Won't retry again` +
+            ` because of a reached limit ${maxRetry}` +
+            ' max retries.'
+        )
+      }
+      version = currentVersion
+    }
+}
+
+async function _createTemplateOrderAndPayments(checkoutOrder, orderDraft) {
+  try {
+    const checkoutPayments = checkoutOrder.paymentInfo?.payments?.map(
+      (p) => p.obj
+    )
+    let addPaymentActions
+    if (checkoutPayments) {
+      const paymentDrafts = checkoutPayments.map(_createPaymentDraft)
+      const paymentCreateResponses = await Promise.all(
+        paymentDrafts.map((paymentDraft) =>
+          apiRoot.payments().post({ body: paymentDraft }).execute()
+        )
+      )
+      addPaymentActions = paymentCreateResponses.map(
+        (paymentCreateResponse) => ({
+          action: 'addPayment',
+          payment: {
+            type: 'payment',
+            id: paymentCreateResponse.body.id,
+          },
+        })
+      )
+    }
+    const { body: templateOrder } = await apiRoot
+      .orders()
+      .importOrder()
+      .post({ body: orderDraft })
+      .execute()
+
+    if (addPaymentActions)
+      await _addPaymentWithRetry(templateOrder, addPaymentActions)
+  } catch (err) {
+    if (!isDuplicateOrderError(err)) {
+      const errMsg =
+        `Unexpected error on creating template order from checkout order with number: ${checkoutOrder.orderNumber}.` +
+        ` Line item: ${JSON.stringify(orderDraft.lineItems)}`
+      throw new VError(err, errMsg)
+    }
+  }
+}
+
+async function _setCheckoutOrderProcessed(checkoutOrder) {
+  const updateActions = [
+    {
+      action: 'setCustomField',
+      name: 'isSubscriptionProcessed',
+      value: true,
+    },
+  ]
+  await apiRoot
+    .orders()
+    .withId({ ID: checkoutOrder.id })
+    .post({
+      body: {
+        actions: updateActions,
+        version: checkoutOrder.version,
+      },
+    })
+    .execute()
+}
+
+async function _processCheckoutOrder(checkoutOrder) {
+  try {
+    const templateOrderDrafts = _generateTemplateOrderDrafts(checkoutOrder)
+
+    await pMap(
+      templateOrderDrafts,
+      async (templateOrderDraft) => {
+        await _createTemplateOrderAndPayments(checkoutOrder, templateOrderDraft)
```

Comment:
Do we retry on retryable errors ? (From concept: `On retriable errors (5xx, 409)`) 🤔

Reply:
I hope we will retry. This feature is built-in in the nodejs sdk. Right now they're creating a PR to address this: https://github.com/commercetools/nodejs/pull/1764

Reply:
For 409 retry: https://github.com/commercetools/commercetools-subscriptions/pull/5/commits/e765d59d40565147b4b87893a81f5d25378514b4

Reply:
It seems we are adding 409 retries only for adding payment infos, not for order import, or am I wrong on my assumption ?

Reply:
You are right. For order import it's `do or do not, there is no retry` because it's creation. If it fails, then the checkout order is skipped and one has to solve it manually. See this point in our concept:

```
We do not check if template-order exists before creating it but we take into account that in case of errors in the previous run it might eventually hit 400 errors (due to the orderNumber duplication).
In case of 400 errors we check its type and skip if it matches orderNumber issue.
On other 400 errors we log and skip.
On retriable errors (5xx, 409) we retry and limit the amount of retries. In case the amount of retries is reached we abort the job and do not set lastStartTs so that it can retry on the next run.
```
![1_1Wp2WU3fz4qUpqEu3QmtZA](https://user-images.githubusercontent.com/803826/156333629-c62b374b-a756-4c16-8fd7-0ccec9ec673a.jpeg)

Reply:
I will check this again once I do the unit tests for error cases, which will come soon.